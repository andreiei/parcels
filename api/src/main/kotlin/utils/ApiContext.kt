package utils

import services.ParcelRepository
import services.RoutesService

data class ApiContext(
    val parcelRepository: ParcelRepository = ParcelRepository(),
    val routesService: RoutesService = RoutesService(),
)
