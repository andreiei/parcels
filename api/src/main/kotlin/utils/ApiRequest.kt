package utils

import com.amazonaws.services.lambda.runtime.events.APIGatewayProxyRequestEvent
import decodeFromString
import failure.ApiException
import kotlinx.serialization.Serializable

@Serializable
data class ApiRequest(
    val body: String?,
    val path: String,
    val httpMethod: String,
) {
    inline fun <reified T> getBody(): T {
        val rawBody: String = body ?: throw ApiException.BadRequestException("Request body is missing")
        return runCatching {
            decodeFromString<T>(rawBody)
        }
            .getOrElse {
                throw ApiException.BadRequestException("Failed to parse request body into ${T::class.simpleName}")
            }
    }

    fun getCustomPath(): String = "$httpMethod $path"

    companion object {
        fun APIGatewayProxyRequestEvent.toApiRequest(): ApiRequest = ApiRequest(
            body = body,
            path = path,
            httpMethod = httpMethod,
        )
    }
}
