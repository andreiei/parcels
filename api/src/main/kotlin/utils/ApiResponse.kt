package utils

import encodeToString
import kotlinx.serialization.Serializable

@Serializable
data class ApiResponse(
    val body: String?,
    val code: Int,
) {
    companion object {
        inline fun <reified T> withContent(body: T, code: Int = 200): ApiResponse =
            ApiResponse(body = encodeToString(body), code = code)

        fun noContent(): ApiResponse = ApiResponse(body = null, code = 204)
    }
}
