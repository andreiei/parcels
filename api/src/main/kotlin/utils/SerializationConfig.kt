import kotlinx.serialization.json.Json
import kotlinx.serialization.serializer

inline fun <reified T> encodeToString(data: T): String = SerializerConfig.json.encodeToString(serializer(), data)
inline fun <reified T> decodeFromString(data: String): T = SerializerConfig.json.decodeFromString(serializer(), data)

open class SerializerConfig {
    companion object {
        val json: Json =
            Json {
                isLenient = true
                prettyPrint = true
                encodeDefaults = true
                ignoreUnknownKeys = true
            }
    }
}
