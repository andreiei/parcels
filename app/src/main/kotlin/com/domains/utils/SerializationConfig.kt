package com.domains.utils

import kotlinx.serialization.json.Json
import kotlinx.serialization.serializer

inline fun <reified T> encodeToString(data: T): String = SerializerConfig.json.encodeToString(serializer(), data)

open class SerializerConfig {
    companion object {
        val json: Json =
            Json {
                isLenient = true
                encodeDefaults = true
                ignoreUnknownKeys = true
                prettyPrint = true
            }
    }
}
