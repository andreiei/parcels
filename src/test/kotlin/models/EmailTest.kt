package models

import io.kotest.core.spec.style.ShouldSpec
import utils.shouldMatchSnapshot

class EmailTest : ShouldSpec({
    should("validate domains") {
        listOf(
            "peter@domain.com",
            "peter@sub-domain.domain.com",
            "@domain.com",
            "peter",
            "peter@",
            "peter@domain..com",
            "peter@-domain.com",
        )
            .associateWith { domain ->
                when (Email.from(domain)) {
                    null -> "❌"
                    else -> "✅"
                }
            }
            .shouldMatchSnapshot(this)
    }
})
