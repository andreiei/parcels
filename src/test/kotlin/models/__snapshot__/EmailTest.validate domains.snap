{
    "peter@domain.com": "✅",
    "peter@sub-domain.domain.com": "✅",
    "@domain.com": "❌",
    "peter": "❌",
    "peter@": "❌",
    "peter@domain..com": "❌",
    "peter@-domain.com": "❌"
}